src/neuroglancer/sliceview/compresso/compresso.hpp

```diff
@@ -0,0 +1,593 @@
+	const size_t sy = header.sy;
+	const size_t sz = header.sz;
+	const size_t voxels = sx * sy * sz;
+	const size_t xstep = header.xstep;
+	const size_t ystep = header.ystep;
+	const size_t zstep = header.zstep;
+
+	if (sx * sy * sz == 0) {
+		return 10;
+	}
+
+	const size_t nx = (sx + xstep - 1) / xstep; // round up
+	const size_t ny = (sy + ystep - 1) / ystep; // round up
+	const size_t nz = (sz + zstep - 1) / zstep; // round up
+	const size_t nblocks = nz * ny * nx;
+
+	size_t window_bytes = (
+		num_bytes 
+			- CompressoHeader::header_size
+			- (header.id_size * sizeof(LABEL))  
+			- (header.value_size * sizeof(WINDOW))
+			- (header.location_size * sizeof(LABEL))
+	);
+	size_t num_condensed_windows = window_bytes / sizeof(WINDOW);
+
+	// allocate memory for all arrays
+	std::vector<LABEL> ids(header.id_size + 1); // +1 to allow vectorized mapping w/ no if statement guarding zero
+	std::vector<WINDOW> window_values(header.value_size);
+	std::vector<LABEL> locations(header.location_size);
+	std::vector<WINDOW> windows(num_condensed_windows);
+
+	size_t iv = CompressoHeader::header_size;
+	for (size_t ix = 0; ix < ids.size() - 1; ix++, iv += sizeof(LABEL)) {
+		ids[ix + 1] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < window_values.size(); ix++, iv += sizeof(WINDOW)) {
+		window_values[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < locations.size(); ix++, iv += sizeof(LABEL)) {
+		locations[ix] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < num_condensed_windows; ix++, iv += sizeof(WINDOW)) {
+		windows[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+
+	windows = run_length_decode_windows<WINDOW>(windows, nblocks);
+
+	bool* boundaries = decode_boundaries<WINDOW>(
+		windows, window_values, 
+		sx, sy, sz, 
+		xstep, ystep, zstep
+	);
+	windows = std::vector<WINDOW>();
+	window_values = std::vector<WINDOW>();
+
+	uint32_t* components = cc3d::connected_components<uint32_t>(
+		boundaries, sx, sy, sz, header.connectivity
+	);
+
+	if (output == NULL) {
+		return 11;
+	}
+
+	decode_nonboundary_labels(components, ids, sx, sy, sz, output);
+	delete[] components;
+	ids = std::vector<LABEL>();
+
+	int err = decode_indeterminate_locations<LABEL>(
+		boundaries, output, locations, 
+		sx, sy, sz,
+		header.connectivity
+	);
+
+	delete[] boundaries;
+
+	// if err is 0, everything was ok
+	return err;
+}
+
+template <>
+int decompress<void,void>(unsigned char* buffer, size_t num_bytes, void* output) {
+	if (!CompressoHeader::valid_header(buffer)) {
+		return 12;
+	}
+
+	CompressoHeader header(buffer);
+
+	bool window8 = (
+		static_cast<int>(header.xstep) * static_cast<int>(header.ystep) * static_cast<int>(header.zstep) <= 8
+	);
+	bool window16 = (
+		static_cast<int>(header.xstep) * static_cast<int>(header.ystep) * static_cast<int>(header.zstep) <= 16
+	);
+	bool window32 = (
+		static_cast<int>(header.xstep) * static_cast<int>(header.ystep) * static_cast<int>(header.zstep) <= 32
+	);
+
+	if (header.data_width == 1) {
```

Comment:
It looks like you could create a separate function templated on LABEL but not WINDOW, so that you can reduce the redundancy here.

Reply:
This is a great suggestion. That code was so ugly.